### PR TITLE
Part 3: Automated Testing for Prompts

### PR DESCRIPTION
**Pull request by imaustink**
- [ ] Link blog article once published.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -15,6 +15,11 @@ services:
     ports:
       - 7233:7233
       - 8233:8233
+    healthcheck:
+      test: ["CMD-SHELL", "curl --fail http://localhost:8233/ || exit 1"]
+      interval: 10s
+      timeout: 30s
+      retries: 3
```
Comment (imaustink) [inline docker-compose.yml:22]: Add health check to ensure Temporal is running.

File: docker-compose.yml [hunk 2]
```diff
@@ -49,9 +54,12 @@ services:
...
     depends_on:
-      - temporal-dev-server
-      - localstack
-      - postgres
+      temporal-dev-server:
+        condition: service_healthy
+      localstack:
+        condition: service_healthy
+      postgres:
+        condition: service_started
```
Comment (imaustink) [inline docker-compose.yml:62]: Wait for Temporal to be healthy.

File: docker-compose.yml [hunk 3]
```diff
@@ -75,9 +83,12 @@ services:
...
```
Comment (imaustink) [inline docker-compose.yml:91]: Wait for Temporal to be healthy.

File: docker-compose.yml [hunk 4]
```diff
@@ -94,3 +105,33 @@ services:
...
+      - DATABASE_TABLE_NAME=vector_db
+      - OPENAI_API_KEY=${OPENAI_API_KEY}
+    volumes:
+      - ./src:/usr/src/app/src
+      - ./test:/usr/src/app/test
+    command: >
+      sh -c '\
+        sleep 10 \
+        && npx nodemon --watch src --ext ts src/test-prompts-worker.ts \
+      '
```
Comment (imaustink) [inline docker-compose.yml:137]: Added worker for `testPromptsWorkflow`.

File: package.json [hunk 5]
```diff
@@ -6,7 +6,8 @@
     "build": "tsc --build",
     "start": "ts-node src/process-documents-worker.ts && ts-node src/invoke-prompt-worker.ts",
     "process-documents": "ts-node src/process-documents-client.ts",
-    "invoke-prompt": "ts-node src/invoke-prompt-client.ts"
+    "invoke-prompt": "ts-node src/invoke-prompt-client.ts",
+    "test-prompts": "ts-node src/test-prompts-client.ts"
```
Comment (imaustink) [inline package.json:10]: Run `testPromptsWorkflow`.

File: src/activities/invoke-prompt-activities.ts [hunk 6]
```diff
@@ -1,23 +1,10 @@
...
-  if (!_gptModel) {
-    _gptModel = new ChatOpenAI({
-      openAIApiKey: OPENAI_API_KEY,
-      temperature: 0,
-      modelName: 'gpt-3.5-turbo'
-    })
-  }
-  return _gptModel
-}
+const getGPTModel = createMemoizedOpenAI();
```
Comment (imaustink) [inline src/activities/invoke-prompt-activities.ts:7]: Moved to central location for sharing.

File: src/activities/invoke-prompt-activities.ts [hunk 7]
```diff
@@ -75,8 +62,11 @@ export async function invokePrompt(input: InvokePromptInput): Promise<InvokeProm
...
     [ 'system', 'You are a friendly, helpful software assistant. Your goal is to help users write CRUD-based software applications using the the Hatchify open-source project in TypeScript.' ],
-    [ 'system', 'You should respond in short paragraphs, using Markdown formatting, seperated with two newlines to keep your responses easily readable.' ],
-    [ 'system', `Here is the documentation that is relevant to the user's query:` + relevantDocumentation.join('\n\n') ],
+    [ 'system', 'You should respond in short paragraphs, using Markdown formatting, separated with two newlines to keep your responses easily readable.' ],
+    [ 'system', 'Whenever possible, use code examples derived from the documentation provided.' ],
+    [ 'system', 'Import references must be included where relevant so that the reader can easily figure out how to import the necessary dependencies.' ],
+    [ 'system', 'Do not use your existing knowledge to determine import references, only use import references as they appear in the relevant documentation for Hatchify' ],
+    [ 'system', `Here is the Hatchify documentation that is relevant to the user's query:` + relevantDocumentation.join('\n\n') ],
```
Comment (imaustink) [inline src/activities/invoke-prompt-activities.ts:69]: Updated prompt to improve quality of results.

File: src/activities/process-documents-activities.ts [hunk 8]
```diff
@@ -145,33 +145,28 @@ export async function processDocuments(input: ProcessDocumentsInput): Promise<Pr
...
-        connectionString: DATABASE_CONNECTION_STRING
-      } as PoolConfig,
-      tableName: DATABASE_TABLE_NAME,
-      columns: {
-        idColumnName: 'id',
-        vectorColumnName: 'vector',
-        contentColumnName: 'content',
-        metadataColumnName: 'metadata',
-      }
-    }
```
Comment (imaustink) [inline src/activities/process-documents-activities.ts:168]: We close the connection after each execution so memoization leads to an error on the second execution.

File: src/workflows.ts [hunk 9]
```diff
@@ -1,20 +1,16 @@
-import { proxyActivities, uuid4 } from '@temporalio/workflow';
-import * as activities from './activities';
+import { proxyActivities, startChild, uuid4 } from '@temporalio/workflow'
+import * as activities from './activities'
 
-const { createS3Bucket, deleteS3Object, deleteS3Bucket } = proxyActivities<typeof activities>({
+const { createS3Bucket, deleteS3Object, deleteS3Bucket, generatePrompt, invokePrompt, loadTestCases } = proxyActivities<typeof activities>({
```
Comment (imaustink) [inline src/workflows.ts:4]: Consolidated "1 minute" activities.

File: README.md [hunk 10]
```diff
@@ -40,3 +40,19 @@ Generated embeddings are stored in a Postgres table:
...
+npm run invoke-prompt <embeddings workflowID>
+```
+
+## More info
+
+This repo was created for demonstrating concepts outlined in the following articles.
+
+- [Your Next AI Startup Should Be Built on Temporal - Part 1](https://www.bitovi.com/blog/your-next-ai-startup-should-be-built-on-temporal-part-1-document-processing)
+- [Your Next AI Startup Should Be Built on Temporal - Part 2](https://www.bitovi.com/blog/your-next-ai-startup-should-be-built-on-temporal-part-2-prompt-engineering)
+- [Your Next AI Startup Should Be Built on Temporal - Part 3]()
```
Comment (imaustink) [inline README.md:56]: TODO: Update the href to point to the published article

File: src/activities/invoke-prompt-activities.ts [hunk 11]
```diff
@@ -75,8 +62,11 @@ export async function invokePrompt(input: InvokePromptInput): Promise<InvokeProm
 
   const response = await gptModel.invoke([
     [ 'system', 'You are a friendly, helpful software assistant. Your goal is to help users write CRUD-based software applications using the the Hatchify open-source project in TypeScript.' ],
-    [ 'system', 'You should respond in short paragraphs, using Markdown formatting, seperated with two newlines to keep your responses easily readable.' ],
-    [ 'system', `Here is the documentation that is relevant to the user's query:` + relevantDocumentation.join('\n\n') ],
+    [ 'system', 'You should respond in short paragraphs, using Markdown formatting, separated with two newlines to keep your responses easily readable.' ],
+    [ 'system', 'Whenever possible, use code examples derived from the documentation provided.' ],
+    [ 'system', 'Import references must be included where relevant so that the reader can easily figure out how to import the necessary dependencies.' ],
+    [ 'system', 'Do not use your existing knowledge to determine import references, only use import references as they appear in the relevant documentation for Hatchify' ],
```
Comment (phillipskevin) [inline src/activities/invoke-prompt-activities.ts:68]: 😂

File: src/activities/test-prompts-activities.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+  query: string
+  answer: string
+  score: number
+  reason: string
+}
+export async function validateQueryResult(input: ValidateQueryResultInput): Promise<ValidateQueryResultOutput> {
+  const gptModel = getGPTModel()
+
+  const response = await gptModel.invoke([
+    [ 'system', 'You are responsible for verifying that prompt provided invoked the correct response from an LLM. Consider the following question and expected answer as well as the prompt\'s answered.' ],
```
Comment (phillipskevin) [inline src/activities/test-prompts-activities.ts:31]: should `prompt's answered` be `prompt's answer`? It likely doesn't make a difference, just curious.

**Review by phillipskevin — APPROVED** [review]
This looks good! Thank you for the detailed comments on the PR. I just left one comment on a possible typo. Good to merge either way.

**Pull request merged**